spark/src/main/scala/io/substrait/spark/logical/ToLogicalPlan.scala

```diff
@@ -157,11 +156,16 @@ class ToLogicalPlan(spark: SparkSession) extends DefaultRelVisitor[LogicalPlan]
   }
   override def visit(fetch: relation.Fetch): LogicalPlan = {
     val child = fetch.getInput.accept(this)
-    val limit = Literal(fetch.getCount.getAsLong.intValue(), IntegerType)
-    fetch.getOffset match {
-      case 1L => GlobalLimit(limitExpr = limit, child = child)
-      case -1L => LocalLimit(limitExpr = limit, child = child)
-      case _ => visitFallback(fetch)
+    if (fetch.getCount.isPresent) {
+      val limit = Literal(fetch.getCount.getAsLong.intValue(), IntegerType)
+      fetch.getOffset match {
+        case 1L => GlobalLimit(limitExpr = limit, child = child)
+        case -1L => LocalLimit(limitExpr = limit, child = child)
+        case _ => visitFallback(fetch)
```

Comment:
Can Spark not handle the case where both the `offset` and `count` are set?

Reply:
`OFFSET` is supported as of [Spark 3.4](https://spark.apache.org/releases/spark-release-3-4-0.html#features).

Reply:
> Can Spark not handle the case where both the offset and count are set?

Yes, the unit test in this PR sets `limit` and `offset`.   Spark handles `limit` by breaking it into two logical relations `LocalLimit` and `GlobalLimit` (there's an explanation [here](https://towardsdatascience.com/stop-using-the-limit-clause-wrong-with-spark-646e328774f5)).  

In the case of the test query in this PR:

```sql
select l_partkey from lineitem where l_shipdate < date '1998-01-01' " +
        "order by l_shipdate asc, l_discount desc limit 100 offset 1000
```

the spark logical plan is:

```
    GlobalLimit 100
    +- Offset 1000
       +- LocalLimit 1100
          +- Project [l_partkey#7997L]
             +- Sort [l_shipdate#8006 ASC NULLS FIRST, l_discount#8002 DESC NULLS LAST], true
                +- Project [l_partkey#7997L, l_shipdate#8006, l_discount#8002]
                   +- Filter (isnotnull(l_shipdate#8006) AND (l_shipdate#8006 < 1998-01-01))
                      +- Relation ...
```

I'm not sure how this would best translate into a Substrait plan.

The handling of the `limit` clause was already in the codebase as inherited from Gluten.  As @Blizzara noted, this PR adds the `offset`.  But I can understand that this whole area needs rethinking.

Reply:
Based on what you've indicated, is converting `FetchRel(count=X, offset = Y, <input>)` into:
```
    GlobalLimit X
    +- Offset Y
       +- LocalLimit X + Y
          + <converted input>
```
viable?

Generally speaking, Substrait consumers should be written in such a way that they accept all valid Substrait plans, even those not built specifically by or for Spark in this case. It's perfectly valid to set both fields of the FetchRel, even if Spark doesn't.

Reply:
> viable?

Could be... I'll do some more testing to get a better understanding of this.

> Generally speaking, ...

Completely agree.  One of the weaknesses of the current test suite it that is relies entirely on round-tripping the query plans.  I need to improve on this, somehow.

Reply:
From reading the Spark PRs here is my understanding.

When converting from Substrait -> Spark:

```
Fetch(limit>=0, offset>0) => LocalLimit -> GlobalLimitAndOffset
Fetch(limit<0, offset>0) => Offset
Fetch(limit>=0, offset==0) => LocalLimit -> GlobalLimit
```

When converting from Spark -> Substrait then the following logic should work:

```
if current_node == LocalLimit:
  if next_node == GlobalLimitAndOffset:
    if current_node.limit < next_node.limit:
      raise Exception("this plan cannot be converted into Substrait")
    else:
      yield Fetch(limit=next_node.limit, offset=next_node.offset)
      skip(next_node)
  elif next_node == GlobalLimit:
    if current_node.limit < next_node.limit:
      raise Exception("this plan cannot be converted into Substrait")
    else:
      yield Fetch(limit=next_node.limit, offset=0)
      skip(next_node)
  else:
    raise Exception("this plan cannot be converted into Substrait")

# This cannot have been preceded by LocalLimit because we would have
# skipped it above so it is just a global limit
if current_node == GlobalLimit:
  yield Fetch(limit=current_node.limit, offset=0)

if current_node == Offset:
  yield Fetch(limit=-1, offset=current_node.offset)

if current_node == GlobalLimitAndOffset:
  yield Fetch(limit=current_node.limit, offset=current_node.offset)
```

The above logic should always round trip successfully (unless I made a mistake in my understanding)

The case where there is either a `LocalLimit` alone or a `LocalLimit` followed by `GlobalLimit` or `GlobalLimitAndOffset` but the `LocalLimit`'s limit is smaller cannot be supported today.

This plan implements a per-group limit and the closest equivalent Substrait plan would be something like...

```
┌──────┐  ┌───────────────────────────┐  ┌───────┐          ┌───────────┐
│ SCAN │  │ FILTER (partition_id = 1) │  │ LIMIT │ ────┬────│ UNION ALL │
└──────┘  └───────────────────────────┘  └───────┘     │    └───────────┘
                                                       │                 
                                                       │                 
┌──────┐  ┌───────────────────────────┐  ┌───────┐     │                 
│ SCAN │  │ FILTER (partition_id = 2) │  │ LIMIT │ ────┼                 
└──────┘  └───────────────────────────┘  └───────┘     │                 
                                                       │                 
                                                       │                 
                                                       │                 
                  ...                                  │                 
                                                       │                 
                                                       │                 
                                                       │                 
┌──────┐  ┌───────────────────────────┐  ┌───────┐     │                 
│ SCAN │  │ FILTER (partition_id = N) │  │ LIMIT │ ────┘                 
└──────┘  └───────────────────────────┘  └───────┘                       
```

Reply:
Thank you @westonpace, that's really helpful.  I think I've captured that in the amended commit, and added a couple of extra tests